table/src/main/java/org/enso/table/data/column/operation/map/text/StringBooleanOp.java

```diff
@@ -0,0 +1,96 @@
+package org.enso.table.data.column.operation.map.text;
+
+import java.util.BitSet;
+import org.enso.table.data.column.operation.map.MapOperation;
+import org.enso.table.data.column.storage.BoolStorage;
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.data.column.storage.StringStorage;
+import org.enso.table.error.UnexpectedTypeException;
+
+public abstract class StringBooleanOp extends MapOperation<StringStorage> {
+  public StringBooleanOp(String name) {
+    super(name);
+  }
+
+  protected abstract boolean doString(String a, String b);
+
+  protected boolean doObject(String a, Object o) {
+    throw new UnexpectedTypeException("a Text");
+  }
+
+  @Override
+  public Storage runMap(StringStorage storage, Object arg) {
+    // TODO [RW, MK] is it ok to handle null like this? Maybe other storages should do the same? It
```

Comment:
I see no reason to do it TBH – leave the "what to do with nulls" decision to the subclass implementer.

Reply:
That's kind of what I was asking.

The way it is currently implemented (here and in other variants as I did it in the same way as others), if the argument is `null`, the result is always `null` too, i.e. the subclass implemeneter cannot override the `null` behaviour (at least without overriding, that is - reimplementing the whole `runMap`). But at least that way is consistent for both zip and map.

My doubts arose from the `DoubleBooleanOp` being a bit inconsistent - its `map` implementation calls `doObject` for `null`, and so does `zip` variant if called with `ObjectStorage`, but with `DoubleStorage` it just returns `null`s - i.e. if someone implements `doObject` in such a way that `doObject(x, null) != null`, its behaviour will be different for the same data, depending on the manner that data is stored.

Shouldn't this be kept consistent between the datatypes? I see the following possibilities for making it consistent:
1. make `StringBooleanOp` always call `doObject` if the second argument is null (but then also I think `doObject` should be called by `DoubleBooleanOp` for the case when two Double Storages are zipped and the second argument is `null` - currently this returns `null` for these rows and that is inconsistent with allowing the implementer to decide behaviour; but that gets more complicated because currently if the first argument is `null` we also always return null - so the operations can also be asymmetric;
2. make all ops behave like `StringBooleanOp`, that is if either first or second argument is `null`, return `null` for such row;
3. make all ops call `doObject` if any of {first,second} arguments is null - this would be imo a more consistent approach than what is described in (1); but it is a bit more costly - if one of the rows is null but the other is not, we need to box one value and this can increase processing cost greatly for columns handling primitive values;
- similar to above, but instead of calling `doObject` if one of the args is null, have variants like `doLeftNull(double right)`, `doLeftNullObject(Object right)`, `doRightNull(double left)` - this would avoid unnecessary boxing while *truly* allowing the implementer to handle nulls whatever they like; by default they would all return `null` thus being equivalent to (2) but allowing the flexibility.

Reply:
I don't think there is a unified solution and I don't think there is a need for a unified solution. These subclasses are all sort of arbitrary helpers that are inherently ad-hoc, abstracting away certain common patterns. So I would say a null-handling approach is correct if it works nicely with the subclasses of the helper (existing  + potential with a much smaller weight).